### PR TITLE
Beatmap conversion

### PR DESCRIPTION
**Pull request by smoogipoo**
Depends on https://github.com/ppy/osu/pull/465

### CODE REVIEW

File: osu.Game/Beatmaps/Beatmap.cs [hunk 1]
```diff
@@ -58,7 +69,27 @@ public ControlPoint TimingPointAt(double time, out ControlPoint overridePoint)
...
         public double CalculateStarDifficulty() => Ruleset.GetRuleset(BeatmapInfo.Mode).CreateDifficultyCalculator(this).Calculate();
+
+        /// <summary>
+        /// Converts this Beatmap to a <see cref="Beatmap{T}"/> containing another type of <see cref="HitObject"/>.
+        /// </summary>
+        /// <typeparam name="T">The type of HitObject the new Beatmap should contain.</typeparam>
+        /// <returns></returns>
+        public Beatmap<T> ConvertTo<T>(PlayMode playMode) where T : HitObject
+        {
+            return Ruleset.GetRuleset(playMode).CreateBeatmapConverter<T>().Convert(this);
```
Comment (peppy) [inline osu.Game/Beatmaps/Beatmap.cs]: not a huge fan of the generic + playmode argument. hopefully we can drop one.

File: osu.Game/Beatmaps/BeatmapBase.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+{
+    /// <summary>
+    /// Contains basic beatmap properties.
+    /// <para>
+    /// This allows converted beatmaps (<see cref="Beatmap{T}"/>) to refer to an original "base" Beatmap
+    /// for properties that shouldn't change unless in exceptional circumstances.
+    /// Properties here may be set to be overriden in such exceptional cases.
+    /// </para>
+    /// </summary>
+    public class BeatmapBase
```
Comment (peppy) [inline osu.Game/Beatmaps/BeatmapBase.cs]: just call this Beatmap?

File: osu.Game/Beatmaps/BeatmapBase.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        public BeatmapBase(BeatmapBase original = null)
+        {
+            this.original = original;
+        }
+
+        private BeatmapInfo beatmapInfo;
+        public BeatmapInfo BeatmapInfo
+        {
+            get { return beatmapInfo ?? original?.BeatmapInfo; }
+            set { beatmapInfo = value; }
```
Comment (peppy) [inline osu.Game/Beatmaps/BeatmapBase.cs]: how would this even be used? do you have a use case?

File: osu.Game/Beatmaps/BeatmapBase.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            get { return controlPoints ?? original?.ControlPoints; }
+            set { controlPoints = value; }
+        }
+
+        private List<Color4> comboColors;
+        public List<Color4> ComboColors
+        {
+            get { return comboColors ?? original?.ComboColors; }
+            set { comboColors = value; }
```
Comment (peppy) [inline osu.Game/Beatmaps/BeatmapBase.cs]: again, why is this needed?

**Comment by peppy** [comment]
Not a huge fan of moving hitobjects out of the beatmap class. `Beatmap` should probably be `Beatmap : Beatmap<HitObject>`

File: osu.Game/Beatmaps/Beatmap.cs (hunk 1 above)
Comment (smoogipooo) [inline osu.Game/Beatmaps/Beatmap.cs]: Playmode is needed for obvious reasons. Generic is needed for type safety, otherwise what to return? Generic Beatmap and then cast to Beatmap<T> inside hitrenderers? The other option is to not put this in Beatmap, and just leave it in Ruleset, which might be okay since this is non-destructive anyway.

Comment (smoogipooo) [inline osu.Game/Beatmaps/Beatmap.cs]: Uh, cast to `Beatmap<T>` inside hitrenderers, escaping...

File: osu.Game/Beatmaps/BeatmapBase.cs (hunk 3 above)
Comment (smoogipooo) [inline osu.Game/Beatmaps/BeatmapBase.cs]: Taiko rewriting the ControlPoints list during Convert. Unless you mean specifically BeatmapInfo?

Comment (peppy) [inline osu.Game/Beatmaps/BeatmapBase.cs]: convert should probably make a complete copy IMO, and not have a reference to the original beatmap

Comment (peppy) [inline osu.Game/Beatmaps/BeatmapBase.cs]: also this is `BeatmapInfo`, not `ControlPoints`?

**Comment by smoogipooo** [comment]
Beatmap is `Beatmap : Beatmap<HitObject>`

File: osu.Game.Modes.Osu/UI/OsuHitRenderer.cs [hunk 5]
```diff
@@ -17,7 +16,7 @@ public OsuHitRenderer(Beatmap beatmap)
         {
         }
 
-        protected override HitObjectConverter<OsuHitObject> Converter => new OsuHitObjectConverter();
+        protected override PlayMode PlayMode => PlayMode.Osu;
```
Comment (peppy) [inline osu.Game.Modes.Osu/UI/OsuHitRenderer.cs]: not a fan of this. playmode enums should not be added to more places in the code base (beacuse they are going to be removed soon)

File: osu.Game.Modes.Taiko/Beatmaps/TaikoBeatmapConverter.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+using osu.Game.Beatmaps;
+using osu.Game.Modes.Taiko.Objects;
+using System.Collections.Generic;
+
+namespace osu.Game.Modes.Taiko.Beatmaps
+{
+    internal class TaikoBeatmapConverter : IBeatmapConverter<TaikoBaseHit>
+    {
+        public Beatmap<TaikoBaseHit> Convert(Beatmap original)
```
Comment (peppy) [inline osu.Game.Modes.Taiko/Beatmaps/TaikoBeatmapConverter.cs]: shouldn't this be `TaikoHitObject`?

File: osu.Game/Beatmaps/Beatmap.cs (hunk 1 above)
Comment (peppy) [inline osu.Game/Beatmaps/Beatmap.cs]: maybe the convert logic should be in the converter class after all (to avoid this stuff).

`var osuBeatmap = osuConverter.Convert(beatmap)`

File: osu.Game.Modes.Taiko/Beatmaps/TaikoBeatmapConverter.cs (hunk 6 above)
Comment (smoogipooo) [inline osu.Game.Modes.Taiko/Beatmaps/TaikoBeatmapConverter.cs]: Nop.
